### DISCUSSION TITLE
why can't ref struct be as generic arguments of ref struct

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6521] [General] [answered]
```
R<R<int>> rr = new R<R<int>>(); //CS0306: The type 'R<int>' may not be used as a type argument

ref struct R<T>
{

}
```

Why can't ```R<int>``` be as generic arguments of ```R<T>```?
I think ```R<R<int>>``` should be allowed.

**Comment by svick** [comment, marked as answer]
Because generic parameters of `ref struct`s are not special. For example, the following code is valid, but wouldn't work if `T` was a `ref struct`:

```c#
ref struct R<T>
{
    static object Box(T x) => x;
}
```

What is needed is something like a `ref struct` constraint, which is being considered in https://github.com/dotnet/csharplang/issues/1148.
